File: Typescript/src/common/Display.ts
import { isNullOrUndefined } from "util";
import Cluster from "../clope/Cluster";
import { ITransactionStore, Transaction, TransactionElement, UID } from "./Typings";

export default class Display {
    private dataSource: ITransactionStore;
    private tableClusters: number[];
    private classesIDs: Array<[TransactionElement, UID]>;
    private clusterOccurences: Array<Map<UID, number>>;
    private columnNumber: number;

    constructor(columnNumber: number, dataSource: ITransactionStore, tableClusters: number[]) {
        this.dataSource = dataSource;
        this.tableClusters = tableClusters;
        this.columnNumber = columnNumber;

        this.classesIDs = dataSource.GetClassesIDs(columnNumber);
        this.clusterOccurences = new Array<Map<UID, number>>();
    }

    public async Out(): Promise<void> {
        await this.GroupBy();

        const sum = new Array<number>(this.classesIDs.length);
        let tempstr = "CLUSTER";
        for (const id of this.classesIDs) {
            tempstr = tempstr + "\t" + id["0"];
        }

        console.log(tempstr);
        for (let i = 0; i < this.clusterOccurences.length; i++) {
            tempstr = (i + 1).toString();
            for (let j = 0; j < this.classesIDs.length; j++) {
                const val = this.clusterOccurences[i].get(this.classesIDs[j]["1"]);
                if (val == null) { throw new Error("Couldnt find value to output"); }
                tempstr += "\t" + val;
                if (sum[j] == null) { sum[j] = 0; }
                sum[j] += val;
            }
            console.log(tempstr);
        }
        console.log("Total");
        tempstr = "";
        for (let j = 0; j < this.classesIDs.length; j++) {
            tempstr += "\t" + sum[j];
        }
        console.log(tempstr);
    }

    private async GroupBy(): Promise<void> {
        let rowNumber = 0;
        await this.dataSource.ReadAll((transaction: Transaction) => {
            const uid = transaction[this.columnNumber];
            const clusterNumber = this.tableClusters[rowNumber++];
            if (clusterNumber >= this.clusterOccurences.length) {
                for (let i = this.clusterOccurences.length; i <= clusterNumber; i++) {
                    this.clusterOccurences[i] = new Map<UID, number>();
                    this.classesIDs.forEach((classElement) => this.clusterOccurences[i].set(classElement["1"], 0));
                }
            }
            const val = this.clusterOccurences[clusterNumber].get(uid);
            if (!isNullOrUndefined(val)) { this.clusterOccurences[clusterNumber].set(uid, val + 1); }
        });
    }
}

File: c#/Clope/common/Display.cs
using System;
using System.Collections.Generic;

namespace Clope.common
{
    public class Display
    {
        private ITransactionStore dataSource;
        private List<int> tableClusters;
        private Dictionary<object, UID> classesIDs;
        private List<Dictionary<UID, int>> clusterOccurences;
        private int columnNumber;

        public Display(int columnNumber, ITransactionStore dataSource, List<int> tableClusters)
        {
            this.dataSource = dataSource;
            this.tableClusters = tableClusters;
            this.columnNumber = columnNumber;

            classesIDs = this.dataSource.GetClassesIDs(columnNumber);
            clusterOccurences = new List<Dictionary<UID, int>>();
        }
        public void Out() {
            GroupBy();

            var sum = new int[classesIDs.Count];
            var tempstr = "CLUSTER";
            foreach (var pair in classesIDs) {
                tempstr = tempstr + "\t" + pair.Key;
            }

            Console.WriteLine(tempstr);
            for (var i = 0; i< this.clusterOccurences.Count; i++) {
                tempstr = (i + 1).ToString();
                var j = 0;
                foreach (var pair in classesIDs)
                {
                    if (clusterOccurences[i].TryGetValue(pair.Value, out var val))
                    {
                        tempstr += "\t" + val;
                        sum[j] += val;
                        j++;
                    }
                }
                //for (var j = 0; j< this.classesIDs.Count; j++)
                //    if (clusterOccurences[i].TryGetValue(classesIDs[j].Value, out var val))
                //    {
                //        tempstr += "\t" + val;
                //        sum[j] += val;
                //    }
                Console.WriteLine(tempstr);
            }
            Console.WriteLine("Total");
            tempstr = "";
            for (var j = 0; j< this.classesIDs.Count; j++) {
                tempstr += "\t" + sum[j];
            }
            Console.WriteLine(tempstr);
        }

        private void GroupBy()
        {
            var rowNumber = 0;
            foreach (Transaction transaction in dataSource) {
                var uid = transaction[columnNumber];
                var clusterNumber = tableClusters[rowNumber++];
                if (clusterNumber >= clusterOccurences.Count)
                    for (var i = clusterOccurences.Count; i <= clusterNumber; i++) {
                        clusterOccurences.Add(new Dictionary<UID, int>());
                        foreach (var pair in classesIDs)
                            clusterOccurences[i].Add(pair.Value, 0);
                    }

                if (clusterOccurences[clusterNumber].TryGetValue(uid, out var val))
                    clusterOccurences[clusterNumber][uid] = val + 1;
            }
        }
    }
}